Fixups for Crowbar Backup/Restore

MaximilianMeister opened the pull request:
https://trello.com/c/9tQEIY2K/573-5-p1-5-finishing-crowbar-backup-restore-upgrade-pr

vuntz commented on crowbar_framework/app/models/backup.rb:
```diff
@@ -114,10 +117,10 @@ def create_archive
 
   def save_archive
-    self.name = file.original_filename.split(".").first
+    self.name = file.original_filename.remove(".tar.gz")
```
Thanks.

This makes me wonder: should we have a validation when uploading a tarball that the extension is `.tar.gz`?

MaximilianMeister replied:
we have that in the frontend at least, this is taken care of by the fileinput.js

rsalevsky replied:
We should also validate the extension in the model. Just adapt this https://github.com/MaximilianMeister/crowbar-core/blob/fixup/crowbar-backup/crowbar_framework/app/models/backup.rb#L34 regex.

MaximilianMeister replied:
isn't it `file.original_filename` rather than `name`? I guess we need a custom validation for the file extension

vuntz commented on crowbar_framework/app/models/backup.rb:
```diff
@@ -23,6 +23,9 @@ class Backup < ActiveRecord::Base
   before_validation :save_or_create_archive, on: :create
   after_validation :delete_archive, unless: -> { errors.empty? }
   after_destroy :delete_archive
+  after_create do
```
I think `after_create` is not good enough as we call `data` possibly at other times (not just when creating) -- for instance, when restoring.

MaximilianMeister replied:
I am a bit lost here. I don't have any idea how to achieve this.
* add the `data.rmtree` to more callbacks (wouldn't help)
* use `ObjectSpace.define_finalizer` (discouraging) see e.g. [here](https://www.ruby-forum.com/topic/17155)

how can we fix this?

MaximilianMeister replied:
hmm this one is still nagging... @vuntz shall I drop it until we have a solution (we should at least track it in a card)? it shouldn't block the other fixes IMO

vuntz commented:
Thanks a lot for the cleanups!

vuntz commented on crowbar_framework/lib/crowbar/backup/restore.rb:
```diff
@@ -36,13 +36,13 @@ def restore
+            if file.basename.to_s.include?("-default.json") && type == :databags
               bc_name = file.basename.to_s.split("-").first
               Proposal.create(barclamp: bc_name, name: "default", properties: json)
               SchemaMigration.run_for_bc(bc_name)
```
Replying to https://github.com/crowbar/crowbar-core/pull/144#discussion_r50018136 here (easier to get the discussion in the new PR):

if you look two lines below here, we do a `record.save`. That's a chef object that is being saved, and that will fail if chef is down. We try to deal with such exceptions elsewhere in the code (`Net::HTTPServerException` or `Errno::ECONNREFUSED`, should not be too hard to find examples).

We should just catch the exception and raise some nice "Restore failed" exception somehow.

MaximilianMeister replied:
thanks for the hint

rsalevsky commented on crowbar_framework/app/models/backup.rb:
```diff
@@ -23,6 +23,9 @@ class Backup < ActiveRecord::Base
   after_validation :delete_archive, unless: -> { errors.empty? }
   after_destroy :delete_archive
+  after_create do
+    data.rmtree
```
This is not needed, Rails will delete the tmpdir when the instance is collected by the garbage collector.

MaximilianMeister replied:
when I checked, my `/tmp` folder has not been cleaned up. when exactly does this garbage collection happen?

rhafer commented on crowbar_framework/lib/crowbar/backup/restore.rb:
```diff
@@ -36,13 +36,13 @@ def restore
-            if file.basename.to_s.include?("-default.json")
+            if file.basename.to_s.include?("-default.json") && type == :databags
               bc_name = file.basename.to_s.split("-").first
               Proposal.create(barclamp: bc_name, name: "default", properties: json)
```
This needs to be ```properties: record```, I guess?

MaximilianMeister replied:
right, how could this ever have worked? :-/

MaximilianMeister commented:
@vuntz the owner/permission issue I would like to handle in another PR, would that be fine with you? i removed the WIP for now on this one

vuntz commented:
> @vuntz the owner/permission issue I would like to handle in another PR, would that be fine with you?

Sure, I'm all for non-monster PRs :-)

MaximilianMeister commented:
rebased. maybe that caused the gating in tempest to fail...

rsalevsky commented on crowbar_framework/lib/crowbar/backup/restore.rb:
```diff
@@ -26,7 +26,7 @@ def initialize(backup)
 
       def restore
-        [:crowbar, :knife, :database].each do |component|
+        [:restore_crowbar, :run_installer, :restore_chef, :restore_database].each do |component|
```
We are already in the restore lib so drop the restore prefix.

MaximilianMeister replied:
I have taken @vuntz suggestion [here](https://github.com/crowbar/crowbar-core/pull/144/files#r49711356)
to me personally it does not really matter

rsalevsky replied:
@tboerger suggested to remove it

vuntz replied:
Should we rename this method to have no name, then? ;-) Note that I don't care too much -- I prefer the new naming, and don't understand what's good in the old one but will live with anything.

rhafer commented on crowbar_framework/lib/crowbar/backup/restore.rb:
```diff
@@ -36,26 +36,32 @@ def restore
+              file = Pathname.new(file)
+              next unless file.extname == ".json"
+              record = JSON.load(file.read)
+              if file.basename.to_s.include?("-default.json") && type == :databags
```
There's a problem with this. Not all proposal end in ```-default.json```. All proposals that can have multiple instances have the instance name here, instead of ```default```. This is especially true for the pacemaker proposals (but there are a few others as well)

MaximilianMeister replied:
right, so the proposals will have a name that the `[user has defined].json`? then we maybe have to figure out the right regex. I will have a look

vuntz replied:
I somehow feel that you can just restore all data bags as proposals when upgrading from Cloud 5. Or is there any data bag that you want to filter out?

rsalevsky replied:
Will this work when you restore from Cloud 6 as well?

vuntz replied:
No, but this code path (moving a proposal from data bag to proposal to database) is specific to the cloud 5 upgrade...

rhafer replied:
Actually there are the network databags (that store the ip allocations) which are not really proposals.

vuntz replied:
Good catch. We can filter them out with a match on `/_network$/`.

MaximilianMeister replied:
> We can filter them out with a match on /_network$/.

so we filter all `_network` and all `bc-template*` ? then the rest is the proposals that we want? would that be fine?

rhafer replied:
Yeah, I think that shoudl work.

rhafer commented on crowbar_framework/lib/crowbar/backup/restore.rb:
```diff
@@ -36,26 +36,32 @@ def restore
+              next unless file.extname == ".json"
+              record = JSON.load(file.read)
+              if file.basename.to_s.include?("-default.json") && type == :databags
+                bc_name = file.basename.to_s.split("-").first
```
Hm, for some reason databags restored from Tex still contain that old ```bc-``` prefix. Isn't that supposed to migrated somewhere? (At least it doesn't seem to work)

MaximilianMeister replied:
this was the PR where the renaming took place https://github.com/crowbar/crowbar-core/pull/46

Not sure about the migration, @vuntz do we have such a thing or do we have to implement it yet for the upgrade case?

rsalevsky replied:
The upgrade should remove this prefix: https://github.com/crowbar/crowbar-core/blob/master/crowbar_framework/lib/crowbar/upgrade.rb#L56-L62

rhafer replied:
Right, it's just not executed. It seems ```supported?``` is returning false (https://github.com/crowbar/crowbar-core/blob/master/crowbar_framework/lib/crowbar/upgrade.rb#L32)

rsalevsky replied:
Oh good catch, @MaximilianMeister can you implement that you abort the restore in case that the version is not the same as the running crowbar version and an upgrade is not possible?

rhafer replied:
Actually the bug seems that ```version``` is a float, while ```supported``` expects it to be a string

rhafer commented on crowbar_framework/lib/crowbar/backup/restore.rb:
```diff
@@ -36,26 +36,32 @@ def restore
+              record = JSON.load(file.read)
+              if file.basename.to_s.include?("-default.json") && type == :databags
+                bc_name = file.basename.to_s.split("-").first
+                Proposal.create(barclamp: bc_name, name: "default", properties: record)
```
This is still not right. I think we need ```record.raw_data``` here. Otherwise the json will include that "json_class" and "chef_type" stuff, which ```Proposal``` doesn't like.

rhafer replied:
I can confirm that ```record.raw_data``` works here.

rhafer commented:
I am fine with merging this in the current state and fixing the remaining issues in subsequent pull request as this brings as a lot closer to a working restore of tex 5 backups. 

+1

MaximilianMeister commented on crowbar_framework/app/models/backup.rb:
```diff
@@ -159,4 +165,10 @@ def validate_hostname
   end
+
+  def validate_upload_file_extension
+    return if file && file.original_filename.match(/tar\.gz$/)
```
that's not correct yet

MaximilianMeister commented:
I have dropped the `after_create` where we remove the tmpdir until we have a better solution, and I fixed the `tar.gz` backend validation which was broken. would be good if we could merge it to proceed with the remaining issues

vuntz commented:
+1

> I have dropped the after_create where we remove the tmpdir until we have a better solution

Should we uncheck the item about that in the card?

MaximilianMeister commented:
> Should we uncheck the item about that in the card?

done, thanks for the hint. can someone merge this? i am not allowed...

rhafer commented:
still +1
